pkg/controller/hibernation/azure_actuator.go

```diff
@@ -0,0 +1,231 @@
+package hibernation
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2020-06-01/compute"
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+
+	corev1 "k8s.io/api/core/v1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	hivev1 "github.com/openshift/hive/pkg/apis/hive/v1"
+	"github.com/openshift/hive/pkg/azureclient"
+	controllerutils "github.com/openshift/hive/pkg/controller/utils"
+)
+
+const (
+	azurePowerStatePrefix  = "PowerState/"
+	azureUnknownPowerState = "unknown"
+)
+
+var (
+	azureRunningStates           = sets.NewString("running")
+	azureStoppedStates           = sets.NewString("stopped", "deallocated")
+	azurePendingStates           = sets.NewString("starting")
+	azureStoppingStates          = sets.NewString("stopping", "deallocating")
+	azureRunningOrPendingStates  = azureRunningStates.Union(azurePendingStates)
+	azureStoppedOrStoppingStates = azureStoppedStates.Union(azureStoppingStates)
+	azureNotRunningStates        = azureStoppedOrStoppingStates.Union(azurePendingStates)
+	azureNotStoppedStates        = azureRunningOrPendingStates.Union(azureStoppingStates)
+)
+
+func init() {
+	RegisterActuator(&azureActuator{azureClientFn: getAzureClient})
+}
+
+type azureActuator struct {
+	// azureClientFn is the function to build an Azure client, here for testing
+	azureClientFn func(*hivev1.ClusterDeployment, client.Client, log.FieldLogger) (azureclient.Client, error)
+}
+
+// CanHandle returns true if the actuator can handle a particular ClusterDeployment
+func (a *azureActuator) CanHandle(cd *hivev1.ClusterDeployment) bool {
+	return cd.Spec.Platform.Azure != nil
+}
+
+// StopMachines will stop machines belonging to the given ClusterDeployment
+func (a *azureActuator) StopMachines(cd *hivev1.ClusterDeployment, c client.Client, logger log.FieldLogger) error {
+	logger = logger.WithField("cloud", "azure")
+	azureClient, err := a.azureClientFn(cd, c, logger)
+	if err != nil {
+		return err
+	}
+	machines, err := listAzureMachines(cd, azureClient, azureRunningOrPendingStates, logger)
+	if err != nil {
+		return err
+	}
+	if len(machines) == 0 {
+		logger.Warning("No machines were found to stop")
```

Comment:
Does this deserve to be a warning instead of an info? This could happen if there are machines that are in the Stopping state.

Reply:
The StopMachines call is currently only done when starting to Hibernate. I would think that if you try to hibernate and there's nothing to stop, that's likely not a good thing.

Reply:
Oh, so the controller only ever makes one attempt to stop and start the machines. For some reason, I had it in my head that only ever reconcile while stopping, the controller would attempt to stop any machines that are not in the stopped or stopping state.